Source/options.h

```diff
@@ -14,18 +16,180 @@ enum class StartUpGameOption {
+	virtual void LoadFromIni(string_view category) = 0;
+	virtual void SaveToIni(string_view category) const = 0;
+
+protected:
+	string_view key;
+	string_view name;
+	string_view description;
+	OptionEntryFlags flags;
+	void NotifyValueChanged();
+
+private:
+	std::function<void()> callback;
+};
+
+class OptionEntryBoolean : public OptionEntryBase {
+public:
+	OptionEntryBoolean(string_view key, OptionEntryFlags flags, string_view name, string_view description, bool defaultValue)
+	    : OptionEntryBase(key, flags, name, description)
+	    , defaultValue(defaultValue)
+	    , value(defaultValue)
+	{
+	}
+	[[nodiscard]] bool GetValue();
+	void SetValue(bool value);
+
+	[[nodiscard]] OptionEntryType GetType() const override;
+	[[nodiscard]] string_view GetValueDescription() const override;
+	void LoadFromIni(string_view category) override;
+	void SaveToIni(string_view category) const override;
+
+private:
+	bool value;
+	bool defaultValue;
+};
+
+class OptionEntryListBase : public OptionEntryBase {
+public:
+	[[nodiscard]] virtual size_t GetListSize() const = 0;
+	[[nodiscard]] virtual string_view GetListDescription(size_t index) const = 0;
+	[[nodiscard]] virtual size_t GetActiveListIndex() const = 0;
+	virtual void SetActiveListIndex(size_t index) = 0;
+
+	[[nodiscard]] OptionEntryType GetType() const override;
+	[[nodiscard]] string_view GetValueDescription() const override;
+
+protected:
+	OptionEntryListBase(string_view key, OptionEntryFlags flags, string_view name, string_view description)
+	    : OptionEntryBase(key, flags, name, description)
+	{
+	}
+};
+
+class OptionEntryEnumBase : public OptionEntryListBase {
+public:
+	void LoadFromIni(string_view category) override;
+	void SaveToIni(string_view category) const override;
+
+	[[nodiscard]] virtual size_t GetListSize() const override;
+	[[nodiscard]] virtual string_view GetListDescription(size_t index) const override;
+	[[nodiscard]] size_t GetActiveListIndex() const override;
+	void SetActiveListIndex(size_t index) override;
+
+protected:
+	OptionEntryEnumBase(string_view key, OptionEntryFlags flags, string_view name, string_view description, int defaultValue)
+	    : OptionEntryListBase(key, flags, name, description)
+	    , defaultValue(defaultValue)
+	    , value(defaultValue)
+	{
+	}
+
+	int GetValueInternal();
+
+	void AddEntry(int value, string_view name);
+
+private:
+	int value;
+	int defaultValue;
+	std::vector<string_view> entryNames;
+	std::vector<int> entryValues;
+};
+
+template <typename T>
+class OptionEntryEnum : public OptionEntryEnumBase {
+public:
+	OptionEntryEnum(string_view key, OptionEntryFlags flags, string_view name, string_view description, T defaultValue, std::initializer_list<std::pair<T, string_view>> entries)
+	    : OptionEntryEnumBase(key, flags, name, description, static_cast<int>(defaultValue))
+	{
+		for (auto entry : entries) {
+			AddEntry(static_cast<int>(entry.first), entry.second);
+		}
+	}
+	T GetValue()

```

Comment:
For wrapper types it is idiomatic in C++ to override `operator*`, i.e.:

```suggestion
	T operator*()
```

Also, shouldn't this method be `const`?

Reply:
Would using something like this make sense for DevilutionX?
- https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-nodiscard.html

Reply:
Yeah, and we have it enabled already I think

Reply:
It doesn't trigger because the function is not marked `const`

Reply:
Missed this method. On the other GetXYZ-Methods I used `const` and `[[nodiscard]]`.
Fixed 🙂